grpc/src/client/load_balancing/test_utils.rs

```diff
@@ -0,0 +1,158 @@
+/*
+ *
+ * Copyright 2025 gRPC authors.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ *
+ */
+
+use crate::client::{
+    load_balancing::{
+        ChannelController, ExternalSubchannel, ForwardingSubchannel, LbState, Subchannel,
+        WorkScheduler,
+    },
+    name_resolution::Address,
+};
+use crate::service::{Message, Request, Response, Service};
+use std::{
+    fmt::Display,
+    hash::{Hash, Hasher},
+    ops::Add,
+    sync::Arc,
+};
+use tokio::{
+    sync::{mpsc, Notify},
+    task::AbortHandle,
+};
+
+pub(crate) struct EmptyMessage {}
+impl Message for EmptyMessage {}
+pub(crate) fn new_request() -> Request {
+    Request::new(Box::pin(tokio_stream::once(
+        Box::new(EmptyMessage {}) as Box<dyn Message>
+    )))
+}
+
+// A test subchannel that forwards connect calls to a channel.
+// This allows tests to verify when a subchannel is asked to connect.
+pub(crate) struct TestSubchannel {
+    address: Address,
+    tx_connect: mpsc::UnboundedSender<TestEvent>,
+}
+
+impl TestSubchannel {
+    fn new(address: Address, tx_connect: mpsc::UnboundedSender<TestEvent>) -> Self {
+        Self {
+            address,
+            tx_connect,
+        }
+    }
+}
+
+impl ForwardingSubchannel for TestSubchannel {
```

Comment:
Not related to this PR: Should `ForwardingSubchannel` be a struct instead of a trait?

In the regular delegating pattern, all the structs implement a single interface. What is the intended use of the `delegate()` method? If we need to get the underlying internal `Subchannel` created by the channel, we could achieve this  by adding an `get_internal()` method on the `Subchannel` trait returns an internal subchannel.

Reply:
We want LB policies in the tree to be able to wrap subchannels and override some of its functionality.

Reply:
To clarify, I was asking about the `Subchannel` being a trait, not the `ForwardingSubchannel` which can be implemented by LB policies.

Reply:
It is the `Subchannel` trait that is returned from methods like `new_subchannel`, `subchannel_update`, `pick` etc. And the implementation for it is provided by the `ExternalSubchannel` struct in the channel, which refers to an `InternalSubchannel` that contains the actual subchannel functionality. In tests, we could use a `TestSubchannel` which would also again implement the `Subchannel` trait.

The `ForwardingSubchannel` need not be implemented by all LB policies. It will be required only by LB policies that want to wrap subchannels. I can imagine this being a struct with a blanket implementation for the `Subchannel` trait. But I'm not able to imagine the `Subchannel` itself being a struct. Maybe we can discuss this if we have time during one of our meetings?